sound/soc/sof/intel/hda-dsp.c

```diff
@@ -696,12 +734,23 @@ int hda_dsp_resume(struct snd_sof_dev *sdev)
 		.state = SOF_DSP_PM_D0,
 		.substate = SOF_HDA_DSP_PM_D0I0,
 	};
+	struct hdac_bus *bus = sof_to_bus(sdev);
 	int ret;
 
 	/* resume from D0I3 */
 	if (sdev->dsp_power_state.state == SOF_DSP_PM_D0) {
 		hda_codec_i915_display_power(sdev, true);
 
+		snd_hdac_bus_init_cmd_io(bus);
+		ret = snd_hdac_ext_bus_link_power_up_all(bus);
```

Comment:
@mrajwa why power up all? And if you do power up all, dont you have to power down the ones that werent used before suspend?

Reply:
Because we have power down all at suspend. If we resume from suspend and some additional stream will be started its link must be ready as our trigger procedure doesn't care about links.

Reply:
@mrajwa streams arent going to be magically started upon resuming. So we should power down the links that werent used before suspend and only power up the ones needed. The new ones that are started after resuming will go through the normal route for powering up the link

Reply:
here's the reference to what we do for the S3 suspend case: 
https://github.com/thesofproject/linux/blob/66957f58a2bd160dd77a86f5d2036201b57261cb/sound/soc/sof/intel/hda-dsp.c#L673

Reply:
I think the logic is right, the links needs to be powered down via LCTL.SPA at S0ix suspend. But @mrajwa , I agree with @ranj063 , we need to mimic the logic in hda_resume() and shut down any links that were not active when we entered suspend. Otherwise S0ix resume will result in more links being powered on than what were powered on at S0ix suspend, and this is not right. But copying the logic in hda_resume() should be sufficient.

Reply:
Hmm why on the resume we do `/* turn off the links that were off before suspend */` What is the point of power gating links on resume? Should that be the part of suspend?

Reply:
@mrajwa whats weird to you? You power up all links upon resume then power down the ones that we not powered up before suspend? Seems totally logical to me